### PR TITLE
Add delete method to Multimap returning void

### PR DESCRIPTION
**Pull request by vmarquez**
addresses https://github.com/hazelcast/hazelcast/issues/2719
This depends on https://github.com/hazelcast/hazelcast-client-protocol/pull/103

### CODE REVIEW

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java [hunk 1]
```diff
@@ -148,6 +149,12 @@ public boolean remove(Object key, Object value) {
         MultiMapRemoveCodec.ResponseParameters resultParameters = MultiMapRemoveCodec.decodeResponse(response);
         return new UnmodifiableLazyList<V>(resultParameters.response, getSerializationService());
     }
+    public void delete(Object key) {
```
Comment (Donnerbart) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java]: Just a general style thing: We have a single newline between each function/block in Hazelcast.

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapDataSerializerHook.java [hunk 2]
```diff
@@ -304,6 +307,12 @@ public IdentifiedDataSerializable createNew(Integer arg) {
                 return new EntrySetResponse();
             }
         };
+        constructors[DELETE] = new ConstructorFunction<Integer, IdentifiedDataSerializable>() {
+            @Override
+            public IdentifiedDataSerializable createNew(Integer arg) {
+                return new DeleteOperation();
+            }
+        };
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapDataSerializerHook.java]: The `DELETE_BACKUP` constructor is missing.

File: hazelcast/src/main/java/com/hazelcast/core/MultiMap.java [hunk 3]
```diff
@@ -130,7 +130,18 @@
...
+    /**
+     *  Deletes all the entries with the given key.
+     * <p/>
+     * <p><b>Warning:</b></p>
+     * This method uses <tt>hashCode</tt> and <tt>equals</tt> of the binary form of
+     * the <tt>key</tt>, not the actual implementations of <tt>hashCode</tt> and <tt>equals</tt>
+     * defined in the <tt>key</tt>'s class.
+     *
+     * @param key   the key of the entry to remove
+     * @returnv void
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/core/MultiMap.java]: I don't think `void` is documented, also typo in `@returnv`

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/DeleteOperation.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+import com.hazelcast.core.EntryEventType;
+import com.hazelcast.multimap.impl.MultiMapContainer;
+import com.hazelcast.multimap.impl.MultiMapDataSerializerHook;
+import com.hazelcast.multimap.impl.MultiMapRecord;
+import com.hazelcast.nio.serialization.Data;
+import com.hazelcast.spi.Operation;
+
+import java.util.Collection;
+
+//TODO: should this not be backup aware? check other Map/remove/Delete
```
Comment (Donnerbart) [inline hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/DeleteOperation.java]: I don't understand this TODO, the `MultiMapBackupAwareOperation` implements `BackupAwareOperation`

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java [hunk 5]
```diff
@@ -148,6 +149,12 @@ public boolean remove(Object key, Object value) {
...
+        checkNotNull(key, NULL_KEY_IS_NOT_ALLOWED);
+
+        Data keyData = toData(key);
+        MultiMapDeleteCodec.encodeRequest(name, keyData, 4L);
```
Comment (Donnerbart) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java]: Magic number `4L`. Should this be the thread ID?

Comment (vbekiaris) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java]: missing the actual invocation here eg `ClientMessage response = invoke(request, keyData);`

Also, for the client-side implementation you need to:
- create a message task on the member side that will handle the invocation from the client (eg look at `MultiMapRemoveMessageTask` which implements `MultiMap.remove(key)`).
- register that message task in `DefaultMessageTaskFactoryProvider`

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java (hunk 1 above)
Comment (vmarquez) [reply hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java]: 👍 fixed,   will do from now on.

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java (hunk 5 above)
Comment (vmarquez) [reply hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java]: Thanks, fixed.

File: hazelcast/src/main/java/com/hazelcast/core/MultiMap.java (hunk 3 above)
Comment (vmarquez) [reply hazelcast/src/main/java/com/hazelcast/core/MultiMap.java]: Thanks, changed.

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapDataSerializerHook.java (hunk 2 above)
Comment (vmarquez) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapDataSerializerHook.java]: Thanks, fixed.

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/DeleteOperation.java (hunk 4 above)
Comment (vmarquez) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/DeleteOperation.java]: Oops, embarrassing note to myself that made it in when I was just playing around and exploring the guts of Hazelcast. :)

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java (hunk 5 above)
Comment (vmarquez) [reply hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java]: Alright, I added the RemoveMessageTask and registered it.

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/DeleteOperation.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (vmarquez) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/operations/DeleteOperation.java]: I've done the shouldBackup similar to how the `map.clear()` operation does it, think that would work?

File: hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java [hunk 6]
```diff
@@ -98,6 +99,15 @@ protected MultiMapResponse removeInternal(Data dataKey) {
         }
     }
 
+    protected void deleteInternal(Data dataKey) {
+        try {
+            Operation operation = new DeleteOperation(name, dataKey, getThreadId());
+            invoke(operation, dataKey);
```
[the remark this answers is not in the document]
Comment (vmarquez) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java]: I wan't sure if it was OK to treat 'DeleteOperation' the same as remove for the statistics, does what I changed look good?

[the remark this answers is not in the document]
Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java]: I believe we need a check if the cluster version is 3.9 or less here and reject the operation in that case. Something like:
```
if (clusterService.getClusterVersion().isLessOrEqual(Versions.V3_9)) {
            throw new IllegalStateException("The operation is not supported when cluster version is less than 3.10");
        }
```
@vbekiaris is this right? And it should be added to `MultiMapDeleteMessageTask` as well, right?

[the remark this answers is not in the document]
Comment (vbekiaris) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java]: Nice spotting @mmedenjak. In the member-side proxy you can use the convenience method `isClusterVersionLessOrEqual` inherited from `AbstractDistributedObject`. Also, it's best to throw an `UnsupportedOperationException`, like for example [in this message task](https://github.com/vbekiaris/hazelcast/blob/a4e3633cdf6a29e4bfbc78455cf6d36872e8d205/hazelcast/src/main/java/com/hazelcast/client/impl/protocol/task/map/MapEventJournalSubscribeTask.java#L51-L54)

[the remark this answers is not in the document]
Comment (vmarquez) [reply hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java]: @vbekiaris @mmedenjak Alright, I've verified the cluster version is greater than 3.9, else throw an Exception.  Thanks again for the guidance.

**Comment by vmarquez** [comment]
@vbekiaris are there any other changes that I should make?  My client-protocol changes have been merged and it looks like a 1.6 has been published.  Let me know if I should ping someone else too.  Thanks again for your assistance.

**Comment by Donnerbart** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
Hi @vmarquez ! There are two checkstyle issues with this PR:
```
hazelcast/src/main/java/com/hazelcast/core/MultiMap.java:54: Total number of methods is 31 (max allowed is 30).
hazelcast/src/main/java/com/hazelcast/multimap/impl/MultiMapProxySupport.java:271: Line is longer than 130 characters (found 149).
```
Can you address those? The method count can be addressed by adding `@SuppressWarnings("checkstyle:methodcount")` on the class level. Also, run `mvn clean install -Pcheckstyle,findbugs -DskipTests` before pushing a new commit to run checkstyle and findbugs locally and see if they are passing.

**Comment by vmarquez** [comment]
@mmedenjak Oh, I was not aware of the checkstyle.  I'll be doing that from now on, thanks.  I have addressed the issues, rebased and pushed.

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by vmarquez** [comment]
@mmedenjak @vbekiaris Just wanted to know if there's more that needs to be done to get this merged, or if I should ping someone else for a review.  Hope I'm not pestering anyone!  After this gets merged I'd like to pick up another ticket/issue too. :)

File: hazelcast/src/main/java/com/hazelcast/core/MultiMap.java [hunk 7]
```diff
@@ -131,6 +132,18 @@
...
+     * <p/>
+     * <p><b>Warning:</b></p>
+     * This method uses <tt>hashCode</tt> and <tt>equals</tt> of the binary form of
+     * the <tt>key</tt>, not the actual implementations of <tt>hashCode</tt> and <tt>equals</tt>
+     * defined in the <tt>key</tt>'s class.
+     *
+     * @param key the key of the entry to remove
+     */
+
+    void delete(Object key);
```
Comment (vbekiaris) [inline hazelcast/src/main/java/com/hazelcast/core/MultiMap.java]: This newline should be below method declaration

File: hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java [hunk 8]
```diff
@@ -149,6 +150,13 @@ public boolean remove(Object key, Object value) {
         return new UnmodifiableLazyList<V>(resultParameters.response, getSerializationService());
     }
 
+    public void delete(Object key) {
```
Comment (vbekiaris) [inline hazelcast-client/src/main/java/com/hazelcast/client/proxy/ClientMultiMapProxy.java]: We should add a test for client-side delete functionality in `com.hazelcast.client.multimap.ClientMultiMapTest`

**Comment by vmarquez** [comment]
@vbekiaris thanks, I have added a client side test.  Happy to make any more changes if necessary, just lmk:

https://github.com/hazelcast/hazelcast/pull/11732/files#diff-b2578561a9cdeb5169e23f257324cf32R224

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
@vmarquez checkstyle fails, can you please run `mvn -Pcheckstyle,findbugs -DskipTests clean install` locally, fix the checkstyle issues and squash the commits? Other than that, it's a 👍  from me.

**Comment by vmarquez** [comment]
@vbekiaris all the checkstyle issues are now fixed locally, and i've rebased.

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vbekiaris** [comment]
@vmarquez great, let's wait for the PR builder to get green. @Donnerbart @mmedenjak anyone up for a second review?

**Comment by Donnerbart** [comment]
run-hazelcast4

**Comment by Holmistr** [comment]
@mmedenjak @vbekiaris  Anything else needed for merging the PR?

**Comment by mmedenjak** [comment]
@vmarquez can you rebase the PR?
@Donnerbart can you check if your review comments have been addressed?

**Comment by vmarquez** [comment]
@mmedenjak I've rebased on the latest from master.

**Comment by vbekiaris** [comment]
run-lab-run

**Comment by vmarquez** [comment]
when I ran `mvn clean install -Pcheckstyle,findbugs -DskipTests` it came back without any issues, not sure why the checks have failed...

**Comment by mmedenjak** [comment]
@vmarquez there was an issue with cloudbees

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by vmarquez** [comment]
@mmedenjak anything else needed?

**Comment by mmedenjak** [comment]
I haven't reviewed the PR so I'll let @vbekiaris and @Donnerbart decide if it is ready to merge

**Comment by vbekiaris** [comment]
@vmarquez merged, many thanks for your contribution!

**Pull request merged**